### PR TITLE
[fix] fixes issues with componentDidUpdate & imagesloaded

### PR DESCRIPTION
**Pull request by yowainwright**
Thanks for this repo @theolampert. 🙏 

----

## Purposed fixes

`componentDidUpdate` causes flickity instance failure. Specifically, I observed the flickity instance load, and then fail on `componentDidUpdate`.

The app I observed this issue in uses node 8 and React 16.4.

#### This PR does the following

- Adds `.nvmrc` with 8.
- Removes `imagesLoaded` method
- Destructures props
- Only uses `imagesloaded` if it's supposed to
- Separates complete flickity reload for cell reload based on logic **"ready"** logic
- Breaks out steps in `.travis.yml` to debug testing easier

----

- **Note:** tests pass locally. 

1__jwainwright_hq-ml-002653____dropbox_me_git-docs_oss_react-flickity-component__zsh_

----

Thanks again! I'm happy to help get this issue fixed. 🎉

### CODE REVIEW

**Comment by theolampert** [comment]
@yowainwright thanks for the PR, I'll take a look at this today and get back to you 👍

**Comment by yaodingyd** [comment]
@yowainwright Could you give me a demo link of this issue with componentDidUpdate? I've been trying to reproduce this but somehow I always misses it..

**Comment by yowainwright** [comment]
@yaodingyd yep. I'll share a video of a Dynamic QA Environment shortly. 

**Note:** This PR,  fixes the issue that`imagesloaded` is never fully [disabled](https://github.com/theolampert/react-flickity-component/blob/master/src/index.js#L22-L34). 

Also, I _think_ this PR could assist in allowing `imagesloaded` to be treeshaken if `disableImagesLoaded` is set to true. 

I'm happy to help assist in anyway I can.

**Comment by yowainwright** [comment]
@yaodingyd here's a [video](https://www.flickr.com/photos/jjwainwright/44845659692/). 

~Thank you!

**Comment by theolampert** [comment]
@yowainwright I think you might need to make the video public.

**Comment by yowainwright** [comment]
@theolampert done! Sorry about that.

File: src/index.js [hunk 1]
```diff
@@ -15,51 +15,57 @@ class FlickityComponent extends Component {
...
-    this.imagesLoaded();
-  }
-
-  imagesLoaded() {
-    if (!this.props.disableImagesLoaded && canUseDOM) {
+          : draggable;
+      if (!disableImagesLoaded) {
+        imagesloaded(
+          this.carousel,
+          function(instance) {
```
Comment (yaodingyd) [inline src/index.js]: `instance` can be removed (not used)

**Comment by yaodingyd** [comment]
OK! I think I finally figure it out. I'll say this issue is a tricky one, for sure.

This issue is introduced at 6553f1b872d8a9e98db04bd7a71de34e9ae6b2eb to fix #34. `imageLoaded` is default to true, and when children are a bunch of pictures, Flickity need to load images first to correctly calculate cell width then position cells. That's what that fix is for.

So the problem introduced here is actually pretty simple: because Flickity directly modifies DOM, React has no idea that DOM is already updated, and still try to reconcile based on last DOM snapshot, which would result in error. This can be triggered as follows:

React-flickity-component's `cDU`fires -> `this.flkty.deactivate()` runs, DOM is updated -> `imagesloaded` callback `this.flkty.activate()` is scheduled, which could restore DOM -> React-flickity-component's parent's `cDU` fires for a async `setState`, React tries to rerender but lost info on DOM -> error happens because React try to clean node that is already removed by Flickity.

@yowainwright Your PR is awesome! Code is very clean and you fixed one of the said issue: if `disableImagesLoaded` is set to true, then there will be no issue. But still if we have `imagesLoaded` enabled, the issue persists: [demo](https://codesandbox.io/s/0xqwljrn2n) (I used your code as flickity component in flickity.js).

I think except adding a user-controlled callback, there's no other straightforward way to fix this: we don't know how long `imagesloaded` would finish. So I propose add another prop like `flickityDidUpdate` and `flickityDidMount`, and user can put their `cDU` or `cDM` code in here. See [example](https://codesandbox.io/s/n3977k123l)

**Comment by yowainwright** [comment]
@yaodingyd I'll remove instance shortly. Thanks for your feedback!

In regards to the last paragraph in this [comment](https://github.com/theolampert/react-flickity-component/pull/49#issuecomment-424446437), async/await (maybe)?

```javascript 

imagesloaded(
  this.carousel,
  async function() {
    await this.flkty.activate();
}.bind(this));

```

**Comment by yaodingyd** [comment]
Yes please!

**Comment by yowainwright** [comment]
@yaodingyd same PR work for you?

**Comment by yaodingyd** [comment]
Yeah, you can just add onto this one

File: src/index.js [hunk 2]
```diff
@@ -15,62 +15,45 @@ class FlickityComponent extends Component {
...
   renderPortal() {
-    if (!this.carousel) {
-      return null;
-    }
+    if (!this.carousel) return null;
     const mountNode = this.carousel.querySelector('.flickity-slider');
-    if (mountNode) {
-      return createPortal(this.props.children, mountNode);
-    }
+    if (mountNode) createPortal(this.props.children, mountNode);
```
[the remark this answers is not in the document]
Comment (yowainwright) [reply src/index.js]: @yaodingyd why is `return` needed?

**Comment by yowainwright** [comment]
@yaodingyd updated

**Comment by theolampert** [comment]
@yowainwright @yaodingyd thanks very much for this! I'll merge and release a new version

**Comment by yowainwright** [comment]
@theolampert @yaodingyd thanks for working on this!

**Pull request merged**